libsql-server/src/http/user/db_factory.rs

```diff
@@ -46,17 +47,27 @@ pub fn namespace_from_headers(
         return Ok(NamespaceName::default());
     }
 
-    let host = headers
-        .get("host")
-        .ok_or_else(|| Error::InvalidHost("missing host header".into()))?
-        .as_bytes();
-    let host_str = std::str::from_utf8(host)
-        .map_err(|_| Error::InvalidHost("host header is not valid UTF-8".into()))?;
+    let result = if let Some(metadata) = headers.get(NAMESPACE_METADATA_KEY) {
+        metadata
+            .to_str()
+            .map_err(|s| Error::InvalidNamespaceBytes(s))
+            .and_then(|ns| NamespaceName::from_string(ns.into()))
+    } else {
+        headers
+            .get("host")
+            .ok_or_else(|| Error::InvalidHost("missing host header".into()))
+            .and_then(|res| Ok(res.as_bytes()))
+            .and_then(|host| {
+                std::str::from_utf8(host)
+                    .map_err(|_| Error::InvalidHost("host header is not valid UTF-8".into()))
+            })
+            .and_then(|host_str| split_namespace(host_str))
+    };
```

Comment:
The behaviour here changed in two ways:
1. If host header is missing then previously we were always returning Err. New version returns Err only if disbale_default_namespace is true
2. If std::str::from_utf8(host) fails then previously we were always returning Err. New version returns Err only if disable_default_namespace is true

I think it is safer to keep the old behaviour. New behaviour is covering error cases like host error being set to wrong value.

Reply:
I agree with it being safer not to fall back to default so aggressively
but we now get into an "exception driven" logic territory
e.g. error out when host header non utf, but fallback to default, when host header malformed
is there a reason for this special treatment of that particular error?

Reply:
Let's then fallback to default only when the header is missing and otherwise just error out with appropriate error. It seems safer and is more in line with the logic that was there before.

Reply:
@haaawk, I restructured that section for backwards compatibility, let me know if you have further comments